buildbot_nix/failed_builds.py

```diff
@@ -0,0 +1,61 @@
+import dbm
+from datetime import datetime
+from pathlib import Path
+from typing import TYPE_CHECKING, Any
+
+from pydantic import BaseModel
+
+if TYPE_CHECKING:
+    database: None | dbm._Database = None
+else:
+    database: Any = None
+
+
+class FailedBuildsError(Exception):
+    pass
+
+
+class FailedBuild(BaseModel):
+    derivation: str
+    time: datetime
+
+
+DB_NOT_INIT_MSG = "Database not initialized"
+
```

Comment:
We can avoid the error above above by adding a database class:

``` 
class Database:
  def __init__(db_path: Path): -> None:
    self.handle = dbm.open(str(db_path), "c")

  def add_build(self, derivation: str, time: datetime) -> None:
    self.handle[derivation] = FailedBuild(derivation=derivation, time=time).model_dump_json()
  # ...
```

Reply:
Looks overall like it would be less code, we just need to pass down the database into the the eval steps where we need it.

Reply:
I wanted to avoid threading it through for now as we plan to put the data into postgres anyway. And then we'd be accessing it through buildbots machinery hopefully. Treating this as a temporary solution so tried to make it as self contained as possible

Reply:
I don't see what you mean by threading?

Reply:
I meant adding it everywhere only to then remove it later when once we integrate with buildbots database we'll access it through `self.master.db` so we'd be needlessly adding arguments and then removing them. That was my initial thinking, but tbh we've spent more time talking about it than I would have spent doing it 😂